router.go

```diff
@@ -36,19 +37,27 @@ type serviceProcedure struct {
 	procedure string
 }
 
+type serviceProcedureEncoding struct {
+	service   string
+	procedure string
+	encoding  transport.Encoding
+}
+
 // MapRouter is a Router that maintains a map of the registered
 // procedures.
 type MapRouter struct {
-	defaultService string
-	entries        map[serviceProcedure]transport.Procedure
+	defaultService            string
+	serviceProcedures         map[serviceProcedure]transport.Procedure
+	serviceProcedureEncodings map[serviceProcedureEncoding]transport.Procedure
```

Comment:
Agree that it’s two look ups in the legacy case, but it’s only one in the common case, and the common case is the only case that matters. It is the only case that will remain when we remove the legacy case and in practice is the only case that will be exercised.